ksqlDB aggresively checks generated schema compatibility in CREATE STREAM statements

ksqlDB converts schemas in schema registry to SQL schemas when a `CREATE` statement is issued without a schema. Then, it generates a `ParsedSchema` from the `SQL` schema and checks that the round-trip schema is compatible with the original schema in schema registry.

This bug was probably introduced as fallout from #4717 and #5756 (which are present in 0.8.0 and 0.10.1 respectively, neither have made it into a version of CP yet as of 5.5.1). Previously, we were not checking the compatibility of schemas until we tried to produce to the topic, now we're checking it up-front - since most source topics aren't ever produced to, this bug was limited to `INSERT INTO` scenarios, but now it prevents users from reading their data.

It may not be feasible to remove this check, because users can declare subset schemas on topics with existing schema registry subjects (e.g. `CREATE STREAM (col3 VARCHAR) WITH (kafka_topic='foo', value_format='avro');` where `foo-value` in schema registry has `col1`, `col2` and `col3`). 

There are an additional class of bugs (see #5673 and discussion) that are caused by mismatched declared schemas and reigstered schemas, so it might be worthwhile to just prevent partial schema inference.

One way this bug manifests itself is that `srClient.testCompatbility` is case-sensitive for `JSON_SR`, so registering the following schema works:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Flight",
  "type": "object",
  "properties": {
    "ICAO24": {
      "type": "string",
      "description": "Unique ICAO 24-bit address of the transponder in hex string representation."
    }
  }
}
```

But changing the field from `ICAO24` to `icao24` will fail with `Could not register schema for subject 'foo-value' because it is incompatible with existing schema.` (ksqlDB capitalizes all fields by default). A targeted fix for the above, for `JSON_SR` schemas, ksql should perserve the case and backtick everything.

Another case where this fails is when the schema compatiblity mode is stricter than `BACKWARDS` - for example `FULL_TRANSITIVE` makes it almost certain that ksql registered schemas will not be compatible with previous schemas because the fields are nullable.

It seems that avro backwards compatibility is much looser than JSON backwards compatibility. Here's an enumeration of things I tried on 5.5.1 and whether or not they worked:

*Register AVRO Schema with lowercase fields*: this works, ksqlDB will be able to read fields and if `INSERT INTO` is used, it successfully registers a new schema with uppercase fields:

```
{"subject":"bar-value","version":1,"id":2,"schema":"{..."fields\":[{\"name\":\"col1\",\"type\":[\"null\",\"int\"]},{\"name\":\"col2\",\"type\":{\"type\":\"enum\",\"name\":\"Suit\",\"symbols\":[\"SPADES\",\"HEARTS\"]}}]}"}
{"subject":"bar-value","version":2,"id":3,"schema":"{..."fields\":[{\"name\":\"COL1\",\"type\":[\"null\",\"int\"],\"default\":null},{\"name\":\"COL2\",\"type\":[\"null\",\"string\"],\"default\":null}]}"}
```

*Register AVRO Schema with Enum Type*: this works, ksqlDB will create a `VARCHAR` field and if `INSERT INTO` is used, it successfully registers a new schema with a varchar field

```
{"subject":"bar-value","version":1,"id":2,"schema":"{..."fields\":[{\"name\":\"col1\",\"type\":[\"null\",\"int\"]},{\"name\":\"col2\",\"type\":{\"type\":\"enum\",\"name\":\"Suit\",\"symbols\":[\"SPADES\",\"HEARTS\"]}}]}"}
{"subject":"bar-value","version":3,"id":8,"schema":"{..."fields\":[{\"name\":\"col2\",\"type\":[\"null\",\"string\"],\"default\":null}]}"}
```

*Register JSON Schema with lowercase fields*: reading the fields work, writing the fields do not work:

```
Caused by: org.apache.kafka.common.errors.SerializationException: Error registering JSON schema: {"type":"object","properties":{"ICAO24":{"connect.index":0,"oneOf":[{"type":"null"},{"type":"string"}]}}}
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Schema being registered is incompatible with an earlier schema; error code: 409
```

These examples make me think that the schema compatibility model of AVRO is much more permissive than JSON_SR. The biggest regression from 5.5.x is the ability to _read_ JSON_SR schemas that are case sensitive. It probably makes sense to fix this as a targeted fix for now for JSON_SR with case sensitivity.

--- a/ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjector.java
+++ b/ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjector.java
@@ -110,7 +110,7 @@ public class SchemaRegisterInjector implements Injector {
       final FormatInfo formatInfo,
       final KsqlConfig config,
       final String statementText,
-      final boolean overwriteExisting
+      final boolean registerIfSchemaExists
   ) {
     final Format format = FormatFactory.of(formatInfo);
     if (!format.supportsSchemaInference()) {
@@ -125,16 +125,7 @@ public class SchemaRegisterInjector implements Injector {
         final ParsedSchema parsedSchema =
             format.toParsedSchema(schema.withoutPseudoAndKeyColsInValue().value(), formatInfo);
 
-        if (!overwriteExisting && srClient.getAllSubjects().contains(subject)) {
-          if (!srClient.testCompatibility(subject, parsedSchema)) {
-            throw new KsqlStatementException(
-                String.format(
-                    "Could not register schema for subject "
-                        + "'%s' because it is incompatible with existing schema.",
-                    subject),
-                statementText);
-          }
-        } else {
+        if (registerIfSchemaExists || !srClient.getAllSubjects().contains(subject)) {
           srClient.register(subject, parsedSchema);
         }
       } catch (IOException | RestClientException e) {

--- a/ksqldb-engine/src/test/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjectorTest.java
+++ b/ksqldb-engine/src/test/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjectorTest.java
@@ -23,6 +23,7 @@ import static org.junit.Assert.assertThrows;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoMoreInteractions;
 import static org.mockito.Mockito.when;
@@ -63,6 +64,7 @@ import io.confluent.ksql.util.PersistentQueryMetadata;
 import java.io.IOException;
 import java.util.HashMap;
 import java.util.Optional;
+import org.apache.avro.Schema;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -167,7 +169,7 @@ public class SchemaRegisterInjectorTest {
   }
 
   @Test
-  public void shouldRegisterSchemaForSchemaRegistryEnabledFormatCreateSource()
+  public void shouldRegisterSchemaForSchemaRegistryEnabledFormatCreateSourceIfSubjectDoesntExist()
       throws IOException, RestClientException {
     // Given:
     givenStatement("CREATE STREAM sink (f1 VARCHAR) WITH (kafka_topic='expectedName', value_format='AVRO', partitions=1);");
@@ -179,6 +181,7 @@ public class SchemaRegisterInjectorTest {
     verify(schemaRegistryClient).register("expectedName-value", AVRO_SCHEMA);
   }
 
+  @SuppressWarnings("deprecation") // make sure deprecated method is not called
   @Test
   public void shouldNotReplaceExistingSchemaForSchemaRegistryEnabledFormatCreateSource()
       throws IOException, RestClientException {
@@ -191,24 +194,8 @@ public class SchemaRegisterInjectorTest {
     injector.inject(statement);
 
     // Then:
-    verify(schemaRegistryClient).getAllSubjects();
-    verify(schemaRegistryClient).testCompatibility(eq("expectedName-value"), any(ParsedSchema.class));
-    verifyNoMoreInteractions(schemaRegistryClient);
-  }
-
-  @Test
-  public void shouldThrowOnExistingSchemaForSchemaRegistryEnabledFormatCreateSourceIncompatible()
-      throws IOException, RestClientException {
-    // Given:
-    givenStatement("CREATE STREAM sink (f1 VARCHAR) WITH (kafka_topic='expectedName', value_format='AVRO', partitions=1);");
-    when(schemaRegistryClient.getAllSubjects()).thenReturn(ImmutableSet.of("expectedName-value"));
-    when(schemaRegistryClient.testCompatibility(eq("expectedName-value"), any(ParsedSchema.class))).thenReturn(false);
-
-    // When:
-    final KsqlStatementException e = assertThrows(KsqlStatementException.class, () -> injector.inject(statement));
-
-    // Then:
-    assertThat(e.getMessage(), containsString("Could not register schema for subject 'expectedName-value' because it is incompatible with existing schema."));
+    verify(schemaRegistryClient, never()).register(any(), any(ParsedSchema.class));
+    verify(schemaRegistryClient, never()).register(any(), any(Schema.class));
   }
 
   @Test

